### ISSUE TITLE
[Core] Enormous file handler and memory leaks on windows

### ISSUE THREAD

**Issue by RBrearton** [ray-project/ray#47459]
### What happened + What you expected to happen

As part of a .cmd script on windows, we're starting up ray with `CALL ray start --head --num-cpus 64`. Afterwards, the script spins up a service that makes use of ray, and stays alive indefinitely.

We noticed that, after moving to ray, our resource usage on the machine massively increased. Most obvious is that Ray seems to leak file handles at an alarming rate - we've reproduced this on a few windows machines, all of which happen to be using 2.33.0. Ray is also leaking memory, albeit more slowly, and we have only had one machine taken down by this.

If it helps, we observed file handler leaks on _windows 10 enterprise version 22H2_, and file handler + memory leaks on _Microsoft Windows Server 2019 Standard 10.0.17763_. It's entirely possible that we would observe memory leaks on the windows enterprise machines, but those machines are running ray with far fewer cores.

## Details

Fortunately, we have extremely detailed system logs on one of the machines that exhibits these issues. I'm very happy to give as much information as required to help solve the issue, but our data is sensitive, so I'll have to clean everything first.

The machine from which the following data was extracted was the windows server machine. This data is associated with a single ray worker, of which there were 64 on that machine.

The dip in resource usage on September 1st at around 2pm came from full restart of ray, required after the system ran out of memory.

For context, our usage of ray was extremely light during this period. I've included the `working_set_peak` plot primarily because it shows when we're actually using ray quite nicely. The little bump on Aug 27th was light usage between midnight and around 5am. It was then nearly completely unused until about 6:30am on Aug 31st.

![image](https://github.com/user-attachments/assets/433e083e-cb3b-4386-b936-ee2ef2ee3d85)
![image](https://github.com/user-attachments/assets/b4ef3785-18c3-4b5b-bb39-40cfef146004)
![image](https://github.com/user-attachments/assets/70f1a71d-7db8-4cd0-ba32-a28032fc7d65)
![image](https://github.com/user-attachments/assets/99294ab5-3003-4ef2-93ad-e4998a8b68cc)

### Versions / Dependencies

2.33.0

### Reproduction script

`ray start --head --num-cpus 64`

### Issue Severity

High: It blocks me from completing my task.

**Comment by anyscalesam** [comment]
@mattip can you take a look here; recommend a quick call with @RBrearton if he is available...

**Comment by RBrearton** [comment]
I'm happy to help any way I can by call. I'll be available 00:30-14:30 PDT tomorrow (I'm guessing you're US west coast - either way, I'll be online 8:30am until 10:30pm UK time).

**Comment by anyscalesam** [comment]
@mattip did you get a chance to sync up with @RBrearton on this? Can you pull me into the meeting invite if y'all find a slot; would love to listen in...

**Comment by mattip** [comment]
We met last week. The problem reproduces even with 10 cpus so I should be able to debug locally. I have an idea for a simpler reproducer, hopefully will try it out soon

**Comment by anyscalesam** [comment]
Terrific - TY for the update @mattip

**Comment by mattip** [comment]
It seems to be connected with putting objects on the store and calling a remote, regardless of whether the remote uses the store and regardless of their size. Using this script:
```
@ray.remote
def f(data1, data2):
    return data1 + data2

data1 = np.ones((3, 3))
data2 = np.ones((3, 3))

if use_put:
    data1_ref = ray.put(data1)
    data2_ref = ray.put(data2)

if call_remote:
    if use_put and use_get:
        futures = [f.remote(data1_ref, data2_ref) for i in range(4)]
    else:
        futures = [f.remote(data1, data2) for i in range(4)]
    print(np.sum(ray.get(futures)))  # 72000000.0

```

I see that

call_remote | use_put | use_get | leak?
---|---|---|---
True | True | True |Yes
True | True | False | Yes
False | --- | --- | No
--- | False | --- | No

**Comment by mattip** [comment]
There are 35 threads in the process that is leaking. Could someone provide a little guidance which of those are connected with checking/managing the object store?

**Comment by anyscalesam** [comment]
@jjyao @rkooo567

**Comment by rkooo567** [comment]
@mattip have you been investigating this issue? I'd love to talk for more guidance

**Comment by mattip** [comment]
@rkooo567 yes let's meet. I sent you a DM on the quansight/anyscale slack channel

**Comment by mattip** [comment]
From what I can tell using this one liner to find open file handles on linux, running the test script on linux does not leak handles.
```
$ for pid in $(ps xa | grep ray | cut -f1 -d" "); \
    do echo -n "$pid "; lsof -a -p $pid | wc -l; done
```

**Comment by jjyao** [comment]
Downgrading to P1 as it only affects Windows.

**Comment by jjyao** [comment]
@mattip Do you know for those leaked fds on windows, are they files, sockets or something else?

**Comment by RBrearton** [comment]
> Downgrading to P1 as it only affects Windows.

I appreciate that the majority of ray users are running ray on linux, but it still seems surprising to me that a memory leak capable of taking down our beefy worker (>200GB of RAM) every other week is anything other than a P0 issue.

We would be serious about scaling up a large amount of compute using ray, but we can't justify this while it leaks so heavily.

Rightly or wrongly, I suspect that a large number of industry users will test ray on Windows. It isn't a great look when a typical ~100 core system leaks ~10,000,000 file handles _per day_.

**Comment by mattip** [comment]
> Do you know for those leaked fds on windows, are they files, sockets or something else?

From what I can tell using windows APIs to check the handles, they seem to be "something else". Threads and processes and others all fall into that class.

**Comment by jjyao** [comment]
@mattip do you mind doing some bisect and see which version of Ray starts to have the leak?

**Comment by jjyao** [comment]
Hi @RBrearton it's an overall P1 issue but it's P0 for @mattip (our Windows expert) to work on.

**Comment by g-lisi** [comment]
Hi, is there any update on this?

**Comment by g-lisi** [comment]
@anyscalesam @mattip @jjyao @rkooo567 can we have a status update on this?
